{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "_HASH_BUCKET_SIZE = 1000\n",
    "\n",
    "categorical_features = ['student_id', 'card_id', 'timeslot', 'trans_type', 'category', 'device_name']\n",
    "\n",
    "continus_features = ['amount', 'remained_amount']\n",
    "\n",
    "df = pd.read_csv('../data/consum_cleaned.csv')\n",
    "df[categorical_features] = df[categorical_features].fillna('')\n",
    "df[continus_features] = df[continus_features].fillna(0)\n",
    "df['card_id'] = df['card_id'].apply(str)\n",
    "_CSV_COLUMNS = df.columns.tolist()\n",
    "\n",
    "TOTAL_NUM = df.shape[0]\n",
    "FIFTEEN_PCT = int(TOTAL_NUM * 0.15)\n",
    "_NUM_EXAMPLES = {\n",
    "    'train': TOTAL_NUM - FIFTEEN_PCT * 2,\n",
    "    'validation': FIFTEEN_PCT,\n",
    "}\n",
    "_ID_SIZE = df.student_id.drop_duplicates().count()\n",
    "_LABEL_NUM = 104\n",
    "\n",
    "def build_model_columns():\n",
    "    \"\"\"Builds a set of wide and deep feature columns.\"\"\"\n",
    "    # Continuous variable columns\n",
    "    cate_columns = []\n",
    "    cate_embedding_columns = []\n",
    "    cate_no_embedding_columns = []\n",
    "    continus_columns = []\n",
    "    embedding_features = []\n",
    "    for col in categorical_features:\n",
    "        coluniquelen = len(df[col].unique())\n",
    "        if coluniquelen > _HASH_BUCKET_SIZE:\n",
    "            print(\"col name :{}\\t\\tcount:{}\".format(col, coluniquelen))\n",
    "            embedding_features.append(col)\n",
    "    for feat in categorical_features:\n",
    "        f = tf.feature_column.categorical_column_with_vocabulary_list(feat, df[feat].unique())\n",
    "        if feat in embedding_features:\n",
    "            f1 = tf.feature_column.embedding_column(f, dimension=100)\n",
    "            cate_embedding_columns.append(f1)\n",
    "        else:\n",
    "            f1 = tf.feature_column.indicator_column(f)\n",
    "            cate_no_embedding_columns.append(f1)\n",
    "        cate_columns.append(f)\n",
    "    for feat in continus_features:\n",
    "        f = tf.feature_column.numeric_column(feat)\n",
    "        continus_columns.append(f)\n",
    "    deep_columns = cate_embedding_columns + cate_no_embedding_columns + continus_columns\n",
    "    return deep_columns\n",
    "\n",
    "\n",
    "def input_fn(data_file, num_epochs, shuffle, batch_size, is_classification=False):\n",
    "    \"\"\"Generate an input function for the Estimator.\"\"\"\n",
    "    assert tf.gfile.Exists(data_file), (\n",
    "        '%s not found. Please make sure you have run census_dataset.py and '\n",
    "        'set the --data_dir argument to the correct path.' % data_file)\n",
    "\n",
    "    def parse_csv(value):\n",
    "        tf.logging.info('Parsing {}'.format(data_file))\n",
    "        record_defaults = [[0.0] if c in continus_features else [''] for c in _CSV_COLUMNS]\n",
    "        columns = tf.decode_csv(value, record_defaults)\n",
    "        features = dict(zip(_CSV_COLUMNS, columns))\n",
    "        labels = features.pop('placei')\n",
    "        if is_classification:\n",
    "            classes = tf.strings.to_number(labels, out_type=tf.int64, name='label_class')\n",
    "            return features, classes\n",
    "        else:\n",
    "            return features, labels\n",
    "\n",
    "    # Extract lines from input files using the Dataset API.\n",
    "    dataset = tf.data.TextLineDataset(data_file)\n",
    "\n",
    "    if shuffle:\n",
    "        dataset = dataset.shuffle(buffer_size=_NUM_EXAMPLES['train'])\n",
    "\n",
    "    dataset = dataset.map(parse_csv, num_parallel_calls=5)\n",
    "\n",
    "    # We call repeat after shuffling, rather than before, to prevent separate\n",
    "    # epochs from blending together.\n",
    "    dataset = dataset.repeat(num_epochs)\n",
    "    dataset = dataset.apply(tf.contrib.data.batch_and_drop_remainder(batch_size))\n",
    "    # dataset = dataset.batch(batch_size)\n",
    "    return dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "col name :student_id\t\tcount:41633\n",
      "col name :card_id\t\tcount:62582\n",
      "INFO:tensorflow:Parsing ../data/tf_train.csv\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'BatchDataset' object has no attribute 'values'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-0ed558b8afda>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0miterator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdataset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmake_one_shot_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mone_element\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_next\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeature_column\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minput_layer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdeep_columns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\u001b[0m in \u001b[0;36minput_layer\u001b[0;34m(features, feature_columns, weight_collections, trainable, cols_to_vars)\u001b[0m\n\u001b[1;32m    275\u001b[0m   \"\"\"\n\u001b[1;32m    276\u001b[0m   return _internal_input_layer(features, feature_columns, weight_collections,\n\u001b[0;32m--> 277\u001b[0;31m                                trainable, cols_to_vars)\n\u001b[0m\u001b[1;32m    278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    279\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\u001b[0m in \u001b[0;36m_internal_input_layer\u001b[0;34m(features, feature_columns, weight_collections, trainable, cols_to_vars, scope)\u001b[0m\n\u001b[1;32m    189\u001b[0m   \u001b[0;31m# is wrapped by a `make_template`.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    190\u001b[0m   with variable_scope.variable_scope(\n\u001b[0;32m--> 191\u001b[0;31m       scope, default_name='input_layer', values=features.values()):\n\u001b[0m\u001b[1;32m    192\u001b[0m     \u001b[0mbuilder\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_LazyBuilder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfeatures\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    193\u001b[0m     \u001b[0moutput_tensors\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'BatchDataset' object has no attribute 'values'"
     ]
    }
   ],
   "source": [
    "deep_columns = build_model_columns()\n",
    "train_file = '../data/tf_train.csv'\n",
    "dataset = input_fn(train_file, 1, False, 10)\n",
    "iterator = dataset.make_one_shot_iterator()\n",
    "one_element = iterator.get_next()\n",
    "v = sess.run(one_element)\n",
    "x = tf.feature_column.input_layer(v[0], deep_columns)\n",
    "sess = tf.InteractiveSession()\n",
    "init_op = tf.global_variables_initializer()\n",
    "init_table = tf.tables_initializer()\n",
    "sess.run(init_op)\n",
    "sess.run(init_table)\n",
    "sess.run(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "ename": "FailedPreconditionError",
     "evalue": "Table already initialized.\n\t [[{{node input_layer/category_indicator/category_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer/category_indicator/category_lookup/hash_table, input_layer/category_indicator/category_lookup/Const, input_layer/category_indicator/category_lookup/ToInt64)]]\n\nCaused by op 'input_layer/category_indicator/category_lookup/hash_table/table_init', defined at:\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 193, in _run_module_as_main\n    \"__main__\", mod_spec)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py\", line 16, in <module>\n    app.launch_new_instance()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/traitlets/config/application.py\", line 658, in launch_instance\n    app.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelapp.py\", line 486, in start\n    self.io_loop.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 127, in start\n    self.asyncio_loop.run_forever()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 422, in run_forever\n    self._run_once()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 1432, in _run_once\n    handle._run()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/events.py\", line 145, in _run\n    self._callback(*self._args)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 117, in _handle_events\n    handler_func(fileobj, events)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 450, in _handle_events\n    self._handle_recv()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 480, in _handle_recv\n    self._run_callback(callback, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 432, in _run_callback\n    callback(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 283, in dispatcher\n    return self.dispatch_shell(stream, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 233, in dispatch_shell\n    handler(stream, idents, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 399, in execute_request\n    user_expressions, allow_stdin)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/ipkernel.py\", line 208, in do_execute\n    res = shell.run_cell(code, store_history=store_history, silent=silent)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/zmqshell.py\", line 537, in run_cell\n    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2662, in run_cell\n    raw_cell, store_history, silent, shell_futures)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2785, in _run_cell\n    interactivity=interactivity, compiler=compiler, result=result)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2903, in run_ast_nodes\n    if self.run_code(code, result):\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2963, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"<ipython-input-12-4040b04c0b50>\", line 1, in <module>\n    x = tf.feature_column.input_layer(one_element[0], deep_columns)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 277, in input_layer\n    trainable, cols_to_vars)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 202, in _internal_input_layer\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3332, in _get_dense_tensor\n    return inputs.get(self)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2263, in get\n    transformed = column._transform_feature(self)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3257, in _transform_feature\n    id_weight_pair = self.categorical_column._get_sparse_tensors(inputs)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3041, in _get_sparse_tensors\n    return _CategoricalColumn.IdWeightPair(inputs.get(self), None)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2263, in get\n    transformed = column._transform_feature(self)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3032, in _transform_feature\n    name='{}_lookup'.format(self.key)).lookup(input_tensor)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 1134, in index_table_from_tensor\n    init, default_value, shared_name=shared_name, name=hash_table_scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 280, in __init__\n    super(HashTable, self).__init__(table_ref, default_value, initializer)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 172, in __init__\n    self._init = initializer.initialize(self)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 381, in initialize\n    table.table_ref, self._keys, self._values, name=scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gen_lookup_ops.py\", line 387, in initialize_table_v2\n    values=values, name=name)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py\", line 787, in _apply_op_helper\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py\", line 488, in new_func\n    return func(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 3272, in create_op\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 1768, in __init__\n    self._traceback = tf_stack.extract_stack()\n\nFailedPreconditionError (see above for traceback): Table already initialized.\n\t [[{{node input_layer/category_indicator/category_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer/category_indicator/category_lookup/hash_table, input_layer/category_indicator/category_lookup/Const, input_layer/category_indicator/category_lookup/ToInt64)]]\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFailedPreconditionError\u001b[0m                   Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1291\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1292\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1293\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mOpError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_run_fn\u001b[0;34m(feed_dict, fetch_list, target_list, options, run_metadata)\u001b[0m\n\u001b[1;32m   1276\u001b[0m       return self._call_tf_sessionrun(\n\u001b[0;32m-> 1277\u001b[0;31m           options, feed_dict, fetch_list, target_list, run_metadata)\n\u001b[0m\u001b[1;32m   1278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_call_tf_sessionrun\u001b[0;34m(self, options, feed_dict, fetch_list, target_list, run_metadata)\u001b[0m\n\u001b[1;32m   1366\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_session\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moptions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeed_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfetch_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtarget_list\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1367\u001b[0;31m         run_metadata)\n\u001b[0m\u001b[1;32m   1368\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFailedPreconditionError\u001b[0m: Table already initialized.\n\t [[{{node input_layer/category_indicator/category_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer/category_indicator/category_lookup/hash_table, input_layer/category_indicator/category_lookup/Const, input_layer/category_indicator/category_lookup/ToInt64)]]",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mFailedPreconditionError\u001b[0m                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-49-c74dd41f1878>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0minit_table\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtables_initializer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0msess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minit_op\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0msess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minit_table\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m    885\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    886\u001b[0m       result = self._run(None, fetches, feed_dict, options_ptr,\n\u001b[0;32m--> 887\u001b[0;31m                          run_metadata_ptr)\n\u001b[0m\u001b[1;32m    888\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0mrun_metadata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    889\u001b[0m         \u001b[0mproto_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTF_GetBuffer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_metadata_ptr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_run\u001b[0;34m(self, handle, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1108\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mfinal_fetches\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mfinal_targets\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mfeed_dict_tensor\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1109\u001b[0m       results = self._do_run(handle, final_targets, final_fetches,\n\u001b[0;32m-> 1110\u001b[0;31m                              feed_dict_tensor, options, run_metadata)\n\u001b[0m\u001b[1;32m   1111\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1112\u001b[0m       \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_run\u001b[0;34m(self, handle, target_list, fetch_list, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1284\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mhandle\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1285\u001b[0m       return self._do_call(_run_fn, feeds, fetches, targets, options,\n\u001b[0;32m-> 1286\u001b[0;31m                            run_metadata)\n\u001b[0m\u001b[1;32m   1287\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1288\u001b[0m       \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_do_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_prun_fn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeeds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfetches\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1306\u001b[0m           self._config.experimental.client_handles_error_formatting):\n\u001b[1;32m   1307\u001b[0m         \u001b[0mmessage\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0merror_interpolation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minterpolate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_graph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1308\u001b[0;31m       \u001b[0;32mraise\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode_def\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmessage\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1309\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1310\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0m_extend_graph\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFailedPreconditionError\u001b[0m: Table already initialized.\n\t [[{{node input_layer/category_indicator/category_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer/category_indicator/category_lookup/hash_table, input_layer/category_indicator/category_lookup/Const, input_layer/category_indicator/category_lookup/ToInt64)]]\n\nCaused by op 'input_layer/category_indicator/category_lookup/hash_table/table_init', defined at:\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 193, in _run_module_as_main\n    \"__main__\", mod_spec)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py\", line 16, in <module>\n    app.launch_new_instance()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/traitlets/config/application.py\", line 658, in launch_instance\n    app.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelapp.py\", line 486, in start\n    self.io_loop.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 127, in start\n    self.asyncio_loop.run_forever()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 422, in run_forever\n    self._run_once()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 1432, in _run_once\n    handle._run()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/events.py\", line 145, in _run\n    self._callback(*self._args)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 117, in _handle_events\n    handler_func(fileobj, events)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 450, in _handle_events\n    self._handle_recv()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 480, in _handle_recv\n    self._run_callback(callback, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 432, in _run_callback\n    callback(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 283, in dispatcher\n    return self.dispatch_shell(stream, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 233, in dispatch_shell\n    handler(stream, idents, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 399, in execute_request\n    user_expressions, allow_stdin)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/ipkernel.py\", line 208, in do_execute\n    res = shell.run_cell(code, store_history=store_history, silent=silent)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/zmqshell.py\", line 537, in run_cell\n    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2662, in run_cell\n    raw_cell, store_history, silent, shell_futures)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2785, in _run_cell\n    interactivity=interactivity, compiler=compiler, result=result)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2903, in run_ast_nodes\n    if self.run_code(code, result):\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2963, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"<ipython-input-12-4040b04c0b50>\", line 1, in <module>\n    x = tf.feature_column.input_layer(one_element[0], deep_columns)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 277, in input_layer\n    trainable, cols_to_vars)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 202, in _internal_input_layer\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3332, in _get_dense_tensor\n    return inputs.get(self)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2263, in get\n    transformed = column._transform_feature(self)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3257, in _transform_feature\n    id_weight_pair = self.categorical_column._get_sparse_tensors(inputs)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3041, in _get_sparse_tensors\n    return _CategoricalColumn.IdWeightPair(inputs.get(self), None)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2263, in get\n    transformed = column._transform_feature(self)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3032, in _transform_feature\n    name='{}_lookup'.format(self.key)).lookup(input_tensor)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 1134, in index_table_from_tensor\n    init, default_value, shared_name=shared_name, name=hash_table_scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 280, in __init__\n    super(HashTable, self).__init__(table_ref, default_value, initializer)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 172, in __init__\n    self._init = initializer.initialize(self)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 381, in initialize\n    table.table_ref, self._keys, self._values, name=scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gen_lookup_ops.py\", line 387, in initialize_table_v2\n    values=values, name=name)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py\", line 787, in _apply_op_helper\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py\", line 488, in new_func\n    return func(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 3272, in create_op\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 1768, in __init__\n    self._traceback = tf_stack.extract_stack()\n\nFailedPreconditionError (see above for traceback): Table already initialized.\n\t [[{{node input_layer/category_indicator/category_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer/category_indicator/category_lookup/hash_table, input_layer/category_indicator/category_lookup/Const, input_layer/category_indicator/category_lookup/ToInt64)]]\n"
     ]
    }
   ],
   "source": [
    "init_op = tf.global_variables_initializer()\n",
    "init_table = tf.tables_initializer()\n",
    "sess.run(init_op)\n",
    "sess.run(init_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.5       , -0.00542771,  0.03267819, -0.06834557, -0.1618337 ,\n",
       "        0.06931791, -0.11821896,  0.03961104,  0.06839292,  0.18072768,\n",
       "        0.13669781,  0.10483644,  0.02830074,  0.01686157,  0.00739236,\n",
       "        0.082309  , -0.00197054, -0.07846905, -0.08892803,  0.09540547,\n",
       "       -0.0158    ,  0.04721189, -0.0667192 , -0.14977056, -0.03364233,\n",
       "        0.0102575 , -0.16489199, -0.03690752, -0.17593393, -0.06212981,\n",
       "        0.04176848, -0.04514671,  0.05780005,  0.13186991,  0.01725243,\n",
       "       -0.05645569, -0.13648224,  0.1867404 , -0.0612515 , -0.01681192,\n",
       "       -0.03353823, -0.17015332,  0.10795977, -0.06483619, -0.00929453,\n",
       "       -0.00234832, -0.07309407,  0.03676496, -0.13697648, -0.1901229 ,\n",
       "        0.07651262,  0.06804652, -0.10970445,  0.04430771, -0.04081658,\n",
       "        0.00203471,  0.03239926, -0.05470664, -0.04123081, -0.10199185,\n",
       "        0.02632782,  0.12340486, -0.05761214,  0.03649587, -0.19984211,\n",
       "       -0.04981123, -0.02424479, -0.08442622, -0.07562018, -0.05227572,\n",
       "       -0.03926931, -0.13392974, -0.01306704,  0.12940106, -0.08643721,\n",
       "       -0.10692816, -0.05213068,  0.06478735,  0.07370042, -0.01055074,\n",
       "       -0.01437129, -0.06013421, -0.09202864,  0.04684785, -0.03067927,\n",
       "       -0.09239265, -0.07546581,  0.00539504,  0.1914995 , -0.00132031,\n",
       "        0.15165292,  0.0578979 ,  0.00854332,  0.00053578,  0.05617227,\n",
       "       -0.05385067, -0.15974385, -0.07343969, -0.12631562, -0.01038217,\n",
       "       -0.02257885,  0.        ,  1.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  1.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        , 24.61      , -0.00532882,\n",
       "       -0.12498251,  0.00682804, -0.02194548, -0.00263519, -0.01226739,\n",
       "       -0.00766118, -0.08976339, -0.05377704,  0.03060708,  0.00645088,\n",
       "        0.12419184, -0.11269188, -0.06878437, -0.08080401, -0.08012417,\n",
       "       -0.0946869 ,  0.00458193,  0.05265949, -0.16413824, -0.04249508,\n",
       "       -0.02029451, -0.02652421, -0.05247387,  0.00089977, -0.01563157,\n",
       "       -0.16795081, -0.05183093, -0.09443539, -0.01889146,  0.12454324,\n",
       "       -0.08386422,  0.08434772,  0.18990861,  0.19225378, -0.0572384 ,\n",
       "       -0.15141545, -0.08706572, -0.01406889,  0.12104432,  0.15612735,\n",
       "       -0.11296292,  0.14117058, -0.05561855, -0.06465085,  0.13599072,\n",
       "       -0.03081623, -0.16783194,  0.08103945, -0.04818087, -0.10232269,\n",
       "       -0.07183413,  0.06121306, -0.15912181,  0.16726199, -0.04106032,\n",
       "       -0.06385505,  0.06233823, -0.18449096,  0.12384861,  0.01468864,\n",
       "       -0.14703374, -0.04012446, -0.02044936, -0.01806933,  0.09429158,\n",
       "        0.06612332,  0.13046248, -0.00826323,  0.01554343, -0.05910769,\n",
       "        0.0684957 , -0.047333  ,  0.04976388, -0.06887227, -0.02451397,\n",
       "        0.09950613,  0.040531  , -0.01818061, -0.02372391, -0.07712926,\n",
       "        0.06433892,  0.16035162, -0.19257526, -0.08396404,  0.05860562,\n",
       "        0.0256356 , -0.11066444,  0.03215661, -0.10570192,  0.19732024,\n",
       "       -0.00163657, -0.0843352 , -0.10735321, -0.05492444, -0.00547413,\n",
       "        0.16337097,  0.15825664,  0.01164415,  0.11058723,  1.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  1.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,\n",
       "        0.        ,  0.        ,  0.        ,  0.        ], dtype=float32)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.set_printoptions(suppress=True) # don't use scientific notation\n",
    "a[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'student_id': array([b'2012059140024', b'2012059100025', b'201322030644',\n",
       "         b'201522020746', b'2014080102005', b'201522060723',\n",
       "         b'201321010125', b'201521037019', b'2012060090018',\n",
       "         b'201522035005'], dtype=object),\n",
       "  'card_id': array([b'175928', b'157006', b'201045', b'326667', b'302391', b'329723',\n",
       "         b'306110', b'331421', b'300824', b'327739'], dtype=object),\n",
       "  'amount': array([ 0.5 , -3.12,  0.1 ,  0.1 ,  0.3 ,  0.6 ,  0.2 ,  0.4 ,  0.6 ,\n",
       "          0.1 ], dtype=float32),\n",
       "  'remained_amount': array([ 24.61,  85.63,  19.97, 120.98,  57.32,  83.36, 499.4 ,  52.7 ,\n",
       "          72.26,  94.53], dtype=float32),\n",
       "  'trans_type': array([b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9', b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9',\n",
       "         b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9', b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9',\n",
       "         b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9', b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9',\n",
       "         b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9', b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9',\n",
       "         b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9', b'POS\\xe6\\xb6\\x88\\xe8\\xb4\\xb9'],\n",
       "        dtype=object),\n",
       "  'category': array([b'bath', b'clothes', b'bath', b'bath', b'bath', b'bath', b'bath',\n",
       "         b'bath', b'bath', b'bath'], dtype=object),\n",
       "  'device_name': array([b'\\xe6\\xb7\\x8b\\xe6\\xb5\\xb4',\n",
       "         b'\\xe6\\xb4\\x97\\xe8\\xa1\\xa3\\xe6\\x88\\xbf',\n",
       "         b'\\xe6\\xb7\\x8b\\xe6\\xb5\\xb4', b'\\xe6\\xb7\\x8b\\xe6\\xb5\\xb4',\n",
       "         b'\\xe6\\xb7\\x8b\\xe6\\xb5\\xb4', b'\\xe6\\xb7\\x8b\\xe6\\xb5\\xb4',\n",
       "         b'\\xe6\\xb7\\x8b\\xe6\\xb5\\xb4', b'\\xe6\\xb7\\x8b\\xe6\\xb5\\xb4',\n",
       "         b'\\xe6\\xb7\\x8b\\xe6\\xb5\\xb4', b'\\xe6\\xb7\\x8b\\xe6\\xb5\\xb4'],\n",
       "        dtype=object),\n",
       "  'timeslot': array([b'sleep', b'sleep', b'sleep', b'sleep', b'sleep', b'sleep',\n",
       "         b'sleep', b'sleep', b'sleep', b'sleep'], dtype=object)},\n",
       " array([b'23', b'10', b'13', b'8', b'24', b'25', b'26', b'4', b'16', b'4'],\n",
       "       dtype=object))"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "ename": "FailedPreconditionError",
     "evalue": "Table not initialized.\n\t [[{{node input_layer_7/student_id_embedding/hash_table_Lookup}} = LookupTableFindV2[Tin=DT_STRING, Tout=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_7/student_id_embedding/student_id_lookup/hash_table, input_layer_7/student_id_embedding/to_sparse_input/values, input_layer_7/student_id_embedding/student_id_lookup/hash_table/Const)]]\n\nCaused by op 'input_layer_7/student_id_embedding/hash_table_Lookup', defined at:\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 193, in _run_module_as_main\n    \"__main__\", mod_spec)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py\", line 16, in <module>\n    app.launch_new_instance()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/traitlets/config/application.py\", line 658, in launch_instance\n    app.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelapp.py\", line 486, in start\n    self.io_loop.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 127, in start\n    self.asyncio_loop.run_forever()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 422, in run_forever\n    self._run_once()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 1432, in _run_once\n    handle._run()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/events.py\", line 145, in _run\n    self._callback(*self._args)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 117, in _handle_events\n    handler_func(fileobj, events)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 450, in _handle_events\n    self._handle_recv()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 480, in _handle_recv\n    self._run_callback(callback, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 432, in _run_callback\n    callback(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 283, in dispatcher\n    return self.dispatch_shell(stream, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 233, in dispatch_shell\n    handler(stream, idents, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 399, in execute_request\n    user_expressions, allow_stdin)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/ipkernel.py\", line 208, in do_execute\n    res = shell.run_cell(code, store_history=store_history, silent=silent)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/zmqshell.py\", line 537, in run_cell\n    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2662, in run_cell\n    raw_cell, store_history, silent, shell_futures)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2785, in _run_cell\n    interactivity=interactivity, compiler=compiler, result=result)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2903, in run_ast_nodes\n    if self.run_code(code, result):\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2963, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"<ipython-input-52-d52d341e40b5>\", line 1, in <module>\n    tensor = tf.feature_column.input_layer(v[0], [deep_columns[0]])\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 277, in input_layer\n    trainable, cols_to_vars)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 202, in _internal_input_layer\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2620, in _get_dense_tensor\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2580, in _get_dense_tensor_internal\n    inputs, weight_collections=weight_collections, trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3041, in _get_sparse_tensors\n    return _CategoricalColumn.IdWeightPair(inputs.get(self), None)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2263, in get\n    transformed = column._transform_feature(self)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3032, in _transform_feature\n    name='{}_lookup'.format(self.key)).lookup(input_tensor)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 230, in lookup\n    self._table_ref, key_tensor, self._default_value, name=scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gen_lookup_ops.py\", line 600, in lookup_table_find_v2\n    default_value=default_value, name=name)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py\", line 787, in _apply_op_helper\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py\", line 488, in new_func\n    return func(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 3272, in create_op\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 1768, in __init__\n    self._traceback = tf_stack.extract_stack()\n\nFailedPreconditionError (see above for traceback): Table not initialized.\n\t [[{{node input_layer_7/student_id_embedding/hash_table_Lookup}} = LookupTableFindV2[Tin=DT_STRING, Tout=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_7/student_id_embedding/student_id_lookup/hash_table, input_layer_7/student_id_embedding/to_sparse_input/values, input_layer_7/student_id_embedding/student_id_lookup/hash_table/Const)]]\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFailedPreconditionError\u001b[0m                   Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1291\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1292\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1293\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mOpError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_run_fn\u001b[0;34m(feed_dict, fetch_list, target_list, options, run_metadata)\u001b[0m\n\u001b[1;32m   1276\u001b[0m       return self._call_tf_sessionrun(\n\u001b[0;32m-> 1277\u001b[0;31m           options, feed_dict, fetch_list, target_list, run_metadata)\n\u001b[0m\u001b[1;32m   1278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_call_tf_sessionrun\u001b[0;34m(self, options, feed_dict, fetch_list, target_list, run_metadata)\u001b[0m\n\u001b[1;32m   1366\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_session\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moptions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeed_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfetch_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtarget_list\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1367\u001b[0;31m         run_metadata)\n\u001b[0m\u001b[1;32m   1368\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFailedPreconditionError\u001b[0m: Table not initialized.\n\t [[{{node input_layer_7/student_id_embedding/hash_table_Lookup}} = LookupTableFindV2[Tin=DT_STRING, Tout=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_7/student_id_embedding/student_id_lookup/hash_table, input_layer_7/student_id_embedding/to_sparse_input/values, input_layer_7/student_id_embedding/student_id_lookup/hash_table/Const)]]",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mFailedPreconditionError\u001b[0m                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-52-d52d341e40b5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0msess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minit_op\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# sess.run(init_table)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0msess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtensor\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m    885\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    886\u001b[0m       result = self._run(None, fetches, feed_dict, options_ptr,\n\u001b[0;32m--> 887\u001b[0;31m                          run_metadata_ptr)\n\u001b[0m\u001b[1;32m    888\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0mrun_metadata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    889\u001b[0m         \u001b[0mproto_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTF_GetBuffer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_metadata_ptr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_run\u001b[0;34m(self, handle, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1108\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mfinal_fetches\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mfinal_targets\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mfeed_dict_tensor\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1109\u001b[0m       results = self._do_run(handle, final_targets, final_fetches,\n\u001b[0;32m-> 1110\u001b[0;31m                              feed_dict_tensor, options, run_metadata)\n\u001b[0m\u001b[1;32m   1111\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1112\u001b[0m       \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_run\u001b[0;34m(self, handle, target_list, fetch_list, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1284\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mhandle\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1285\u001b[0m       return self._do_call(_run_fn, feeds, fetches, targets, options,\n\u001b[0;32m-> 1286\u001b[0;31m                            run_metadata)\n\u001b[0m\u001b[1;32m   1287\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1288\u001b[0m       \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_do_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_prun_fn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeeds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfetches\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1306\u001b[0m           self._config.experimental.client_handles_error_formatting):\n\u001b[1;32m   1307\u001b[0m         \u001b[0mmessage\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0merror_interpolation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minterpolate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_graph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1308\u001b[0;31m       \u001b[0;32mraise\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode_def\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmessage\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1309\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1310\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0m_extend_graph\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFailedPreconditionError\u001b[0m: Table not initialized.\n\t [[{{node input_layer_7/student_id_embedding/hash_table_Lookup}} = LookupTableFindV2[Tin=DT_STRING, Tout=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_7/student_id_embedding/student_id_lookup/hash_table, input_layer_7/student_id_embedding/to_sparse_input/values, input_layer_7/student_id_embedding/student_id_lookup/hash_table/Const)]]\n\nCaused by op 'input_layer_7/student_id_embedding/hash_table_Lookup', defined at:\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 193, in _run_module_as_main\n    \"__main__\", mod_spec)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py\", line 16, in <module>\n    app.launch_new_instance()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/traitlets/config/application.py\", line 658, in launch_instance\n    app.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelapp.py\", line 486, in start\n    self.io_loop.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 127, in start\n    self.asyncio_loop.run_forever()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 422, in run_forever\n    self._run_once()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 1432, in _run_once\n    handle._run()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/events.py\", line 145, in _run\n    self._callback(*self._args)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 117, in _handle_events\n    handler_func(fileobj, events)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 450, in _handle_events\n    self._handle_recv()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 480, in _handle_recv\n    self._run_callback(callback, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 432, in _run_callback\n    callback(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 283, in dispatcher\n    return self.dispatch_shell(stream, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 233, in dispatch_shell\n    handler(stream, idents, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 399, in execute_request\n    user_expressions, allow_stdin)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/ipkernel.py\", line 208, in do_execute\n    res = shell.run_cell(code, store_history=store_history, silent=silent)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/zmqshell.py\", line 537, in run_cell\n    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2662, in run_cell\n    raw_cell, store_history, silent, shell_futures)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2785, in _run_cell\n    interactivity=interactivity, compiler=compiler, result=result)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2903, in run_ast_nodes\n    if self.run_code(code, result):\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2963, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"<ipython-input-52-d52d341e40b5>\", line 1, in <module>\n    tensor = tf.feature_column.input_layer(v[0], [deep_columns[0]])\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 277, in input_layer\n    trainable, cols_to_vars)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 202, in _internal_input_layer\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2620, in _get_dense_tensor\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2580, in _get_dense_tensor_internal\n    inputs, weight_collections=weight_collections, trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3041, in _get_sparse_tensors\n    return _CategoricalColumn.IdWeightPair(inputs.get(self), None)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2263, in get\n    transformed = column._transform_feature(self)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3032, in _transform_feature\n    name='{}_lookup'.format(self.key)).lookup(input_tensor)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 230, in lookup\n    self._table_ref, key_tensor, self._default_value, name=scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gen_lookup_ops.py\", line 600, in lookup_table_find_v2\n    default_value=default_value, name=name)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py\", line 787, in _apply_op_helper\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py\", line 488, in new_func\n    return func(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 3272, in create_op\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 1768, in __init__\n    self._traceback = tf_stack.extract_stack()\n\nFailedPreconditionError (see above for traceback): Table not initialized.\n\t [[{{node input_layer_7/student_id_embedding/hash_table_Lookup}} = LookupTableFindV2[Tin=DT_STRING, Tout=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_7/student_id_embedding/student_id_lookup/hash_table, input_layer_7/student_id_embedding/to_sparse_input/values, input_layer_7/student_id_embedding/student_id_lookup/hash_table/Const)]]\n"
     ]
    }
   ],
   "source": [
    "tensor = tf.feature_column.input_layer(v[0], [deep_columns[0]])\n",
    "init_op = tf.global_variables_initializer()\n",
    "init_table = tf.tables_initializer()\n",
    "sess.run(init_op)\n",
    "# sess.run(init_table)\n",
    "sess.run(tensor[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "ename": "FailedPreconditionError",
     "evalue": "Table already initialized.\n\t [[{{node input_layer_3/student_id_embedding/student_id_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_3/student_id_embedding/student_id_lookup/hash_table, input_layer_3/student_id_embedding/student_id_lookup/Const, input_layer_3/student_id_embedding/student_id_lookup/ToInt64)]]\n\nCaused by op 'input_layer_3/student_id_embedding/student_id_lookup/hash_table/table_init', defined at:\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 193, in _run_module_as_main\n    \"__main__\", mod_spec)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py\", line 16, in <module>\n    app.launch_new_instance()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/traitlets/config/application.py\", line 658, in launch_instance\n    app.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelapp.py\", line 486, in start\n    self.io_loop.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 127, in start\n    self.asyncio_loop.run_forever()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 422, in run_forever\n    self._run_once()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 1432, in _run_once\n    handle._run()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/events.py\", line 145, in _run\n    self._callback(*self._args)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 117, in _handle_events\n    handler_func(fileobj, events)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 450, in _handle_events\n    self._handle_recv()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 480, in _handle_recv\n    self._run_callback(callback, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 432, in _run_callback\n    callback(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 283, in dispatcher\n    return self.dispatch_shell(stream, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 233, in dispatch_shell\n    handler(stream, idents, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 399, in execute_request\n    user_expressions, allow_stdin)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/ipkernel.py\", line 208, in do_execute\n    res = shell.run_cell(code, store_history=store_history, silent=silent)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/zmqshell.py\", line 537, in run_cell\n    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2662, in run_cell\n    raw_cell, store_history, silent, shell_futures)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2785, in _run_cell\n    interactivity=interactivity, compiler=compiler, result=result)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2903, in run_ast_nodes\n    if self.run_code(code, result):\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2963, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"<ipython-input-47-4992a01a2421>\", line 1, in <module>\n    tensor = tf.feature_column.input_layer(v[0], [deep_columns[0]])\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 277, in input_layer\n    trainable, cols_to_vars)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 202, in _internal_input_layer\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2620, in _get_dense_tensor\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2580, in _get_dense_tensor_internal\n    inputs, weight_collections=weight_collections, trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3041, in _get_sparse_tensors\n    return _CategoricalColumn.IdWeightPair(inputs.get(self), None)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2263, in get\n    transformed = column._transform_feature(self)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3032, in _transform_feature\n    name='{}_lookup'.format(self.key)).lookup(input_tensor)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 1134, in index_table_from_tensor\n    init, default_value, shared_name=shared_name, name=hash_table_scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 280, in __init__\n    super(HashTable, self).__init__(table_ref, default_value, initializer)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 172, in __init__\n    self._init = initializer.initialize(self)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 381, in initialize\n    table.table_ref, self._keys, self._values, name=scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gen_lookup_ops.py\", line 387, in initialize_table_v2\n    values=values, name=name)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py\", line 787, in _apply_op_helper\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py\", line 488, in new_func\n    return func(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 3272, in create_op\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 1768, in __init__\n    self._traceback = tf_stack.extract_stack()\n\nFailedPreconditionError (see above for traceback): Table already initialized.\n\t [[{{node input_layer_3/student_id_embedding/student_id_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_3/student_id_embedding/student_id_lookup/hash_table, input_layer_3/student_id_embedding/student_id_lookup/Const, input_layer_3/student_id_embedding/student_id_lookup/ToInt64)]]\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFailedPreconditionError\u001b[0m                   Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1291\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1292\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1293\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mOpError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_run_fn\u001b[0;34m(feed_dict, fetch_list, target_list, options, run_metadata)\u001b[0m\n\u001b[1;32m   1276\u001b[0m       return self._call_tf_sessionrun(\n\u001b[0;32m-> 1277\u001b[0;31m           options, feed_dict, fetch_list, target_list, run_metadata)\n\u001b[0m\u001b[1;32m   1278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_call_tf_sessionrun\u001b[0;34m(self, options, feed_dict, fetch_list, target_list, run_metadata)\u001b[0m\n\u001b[1;32m   1366\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_session\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moptions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeed_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfetch_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtarget_list\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1367\u001b[0;31m         run_metadata)\n\u001b[0m\u001b[1;32m   1368\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFailedPreconditionError\u001b[0m: Table already initialized.\n\t [[{{node input_layer_3/student_id_embedding/student_id_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_3/student_id_embedding/student_id_lookup/hash_table, input_layer_3/student_id_embedding/student_id_lookup/Const, input_layer_3/student_id_embedding/student_id_lookup/ToInt64)]]",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mFailedPreconditionError\u001b[0m                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-53-cfe93be9aecf>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minit_table\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(self, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m    885\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    886\u001b[0m       result = self._run(None, fetches, feed_dict, options_ptr,\n\u001b[0;32m--> 887\u001b[0;31m                          run_metadata_ptr)\n\u001b[0m\u001b[1;32m    888\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0mrun_metadata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    889\u001b[0m         \u001b[0mproto_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTF_GetBuffer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_metadata_ptr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_run\u001b[0;34m(self, handle, fetches, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1108\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mfinal_fetches\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mfinal_targets\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mfeed_dict_tensor\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1109\u001b[0m       results = self._do_run(handle, final_targets, final_fetches,\n\u001b[0;32m-> 1110\u001b[0;31m                              feed_dict_tensor, options, run_metadata)\n\u001b[0m\u001b[1;32m   1111\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1112\u001b[0m       \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_run\u001b[0;34m(self, handle, target_list, fetch_list, feed_dict, options, run_metadata)\u001b[0m\n\u001b[1;32m   1284\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mhandle\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1285\u001b[0m       return self._do_call(_run_fn, feeds, fetches, targets, options,\n\u001b[0;32m-> 1286\u001b[0;31m                            run_metadata)\n\u001b[0m\u001b[1;32m   1287\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1288\u001b[0m       \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_do_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_prun_fn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeeds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfetches\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/tensorflow/python/client/session.py\u001b[0m in \u001b[0;36m_do_call\u001b[0;34m(self, fn, *args)\u001b[0m\n\u001b[1;32m   1306\u001b[0m           self._config.experimental.client_handles_error_formatting):\n\u001b[1;32m   1307\u001b[0m         \u001b[0mmessage\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0merror_interpolation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minterpolate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_graph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1308\u001b[0;31m       \u001b[0;32mraise\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode_def\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmessage\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1309\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1310\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0m_extend_graph\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFailedPreconditionError\u001b[0m: Table already initialized.\n\t [[{{node input_layer_3/student_id_embedding/student_id_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_3/student_id_embedding/student_id_lookup/hash_table, input_layer_3/student_id_embedding/student_id_lookup/Const, input_layer_3/student_id_embedding/student_id_lookup/ToInt64)]]\n\nCaused by op 'input_layer_3/student_id_embedding/student_id_lookup/hash_table/table_init', defined at:\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 193, in _run_module_as_main\n    \"__main__\", mod_spec)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py\", line 16, in <module>\n    app.launch_new_instance()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/traitlets/config/application.py\", line 658, in launch_instance\n    app.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelapp.py\", line 486, in start\n    self.io_loop.start()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 127, in start\n    self.asyncio_loop.run_forever()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 422, in run_forever\n    self._run_once()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/base_events.py\", line 1432, in _run_once\n    handle._run()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/asyncio/events.py\", line 145, in _run\n    self._callback(*self._args)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/platform/asyncio.py\", line 117, in _handle_events\n    handler_func(fileobj, events)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 450, in _handle_events\n    self._handle_recv()\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 480, in _handle_recv\n    self._run_callback(callback, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/zmq/eventloop/zmqstream.py\", line 432, in _run_callback\n    callback(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tornado/stack_context.py\", line 276, in null_wrapper\n    return fn(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 283, in dispatcher\n    return self.dispatch_shell(stream, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 233, in dispatch_shell\n    handler(stream, idents, msg)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/kernelbase.py\", line 399, in execute_request\n    user_expressions, allow_stdin)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/ipkernel.py\", line 208, in do_execute\n    res = shell.run_cell(code, store_history=store_history, silent=silent)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/ipykernel/zmqshell.py\", line 537, in run_cell\n    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2662, in run_cell\n    raw_cell, store_history, silent, shell_futures)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2785, in _run_cell\n    interactivity=interactivity, compiler=compiler, result=result)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2903, in run_ast_nodes\n    if self.run_code(code, result):\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2963, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"<ipython-input-47-4992a01a2421>\", line 1, in <module>\n    tensor = tf.feature_column.input_layer(v[0], [deep_columns[0]])\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 277, in input_layer\n    trainable, cols_to_vars)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 202, in _internal_input_layer\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2620, in _get_dense_tensor\n    trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2580, in _get_dense_tensor_internal\n    inputs, weight_collections=weight_collections, trainable=trainable)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3041, in _get_sparse_tensors\n    return _CategoricalColumn.IdWeightPair(inputs.get(self), None)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 2263, in get\n    transformed = column._transform_feature(self)  # pylint: disable=protected-access\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/feature_column/feature_column.py\", line 3032, in _transform_feature\n    name='{}_lookup'.format(self.key)).lookup(input_tensor)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 1134, in index_table_from_tensor\n    init, default_value, shared_name=shared_name, name=hash_table_scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 280, in __init__\n    super(HashTable, self).__init__(table_ref, default_value, initializer)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 172, in __init__\n    self._init = initializer.initialize(self)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/lookup_ops.py\", line 381, in initialize\n    table.table_ref, self._keys, self._values, name=scope)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gen_lookup_ops.py\", line 387, in initialize_table_v2\n    values=values, name=name)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py\", line 787, in _apply_op_helper\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py\", line 488, in new_func\n    return func(*args, **kwargs)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 3272, in create_op\n    op_def=op_def)\n  File \"/home/liwenjie/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py\", line 1768, in __init__\n    self._traceback = tf_stack.extract_stack()\n\nFailedPreconditionError (see above for traceback): Table already initialized.\n\t [[{{node input_layer_3/student_id_embedding/student_id_lookup/hash_table/table_init}} = InitializeTableV2[Tkey=DT_STRING, Tval=DT_INT64, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](input_layer_3/student_id_embedding/student_id_lookup/hash_table, input_layer_3/student_id_embedding/student_id_lookup/Const, input_layer_3/student_id_embedding/student_id_lookup/ToInt64)]]\n"
     ]
    }
   ],
   "source": [
    "sess.run(init_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<tf.Tensor 'input_layer_7/concat:0' shape=(10, 100) dtype=float32>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tensor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
